ast/format.go

```diff
@@ -0,0 +1,301 @@
+package ast
+
+import (
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+)
+
+func Format(n Node) string {
+	return formatNode(n)
+}
+
+func formatChildren(children interface{}, sep string) string {
```

Comment:
Can you update the format implementation to also use the ast.Walk/Visitor paradigm? This way you can avoid the use of reflection and not need to have a psuedo walk implementation as part of the format method.

One way to do this is to use many different visitor types, one for each of the methods that you already have here. 

Additionally can you use the `strings.Builder` type instead of many calls to `fmt.Sprintf`? This will be significantly faster and should help using the walk paradigm as the builder can be passed to children nodes as you walk.

Reply:
Ok, sounds good. I was thinking of it, but I decided to follow an approach similar to JSON. However, what you are saying has much sense to me. Thank you.

Reply:
I spent the day on this, but it seems that visiting to extract the query is quite challenging.

The problem lies in the fact that every node must wait for its children to complete their piece of query before concatenating their results (or wait that they append their result to the StringBuilder).

In order to do this, I added hooks to the visitor in order to be called once a child has finished and also to produce different visitors for each child.  In this way the parent visitor can produce the correct visitor for each children and append pieces of string before and after children do it.

For example, an arrow function needs a visitor that:
 - appends a `(` on visit;
 - lets `Params` append their result using a proper visitor for `PropertyExpression`;
 - appends `)=>`;
 - lets `Body` append its result using a proper visitor for `Node`.

The resulting code becomes much more complex (and less readable) than the one I have already pushed. Plus, the Visitor interface becomes more complex.

I haven't yet tested the current implementation, but, before spending more time on this, do you think it would be better if I simply adapt this implementation to another one that uses a StringBuilder?

If you prefer, I can go on and workout a better solution to use the walk-visitor pattern.